### DISCUSSION TITLE
Firejail versus Flatpak

### DISCUSSION THREAD

**Discussion by CocoR55** [netblue30/firejail#4685] [Q&A] [answered]
Bonjour,

What is the difference between flatpak and firejail?
Both use sandboxes. In which case firejail is preferable and in which case flatpak? 
Which is the least vulnerable?
So, many questions that I have. And you have an idea?
Let me know. 

Regards,
CcoR55

**Comment by rusty-snake** [comment, marked as answer]
Firejail is just a sandboxing solution for programs installed via apt/dnf/pip/cargo/tar.gz/... while flatpak is a full development/building/distribution/running framework. Therefore the first question is "Do I have a choice?" i.e. if a program is only available as flatpak or only as rpm, you have to use flatpak/firejail. If you have a choice, you need to understand the differences between installing via apt/dnf and flatpak (I ignore other ways like pip or tar.gz here). For example are flatpaks always the newest version, if you use a stable distribution and a flatpak comes with a new Gtk version with a theme rework it will not integrate into your DE (visually).

Pro/Con table:

| firejail | flatpak |
| --- | --- |
| finer and more detailed permission model | easier permission model |
| sandbox configuration is maintained by a community who wants to make tight sandboxes (not always possible) | sandbox configuration is maintained by the developer/distributor (often the same person), if he/she does not want to get bug reports like "firefox can not write to ~/.bashrc" the permissions are often very weak (= effectively no sandbox) |
| apps have a shared base system | apps have a private base system |
| dbus filtering | dbus filtering |
| private netns possible | private netns not possible (without custom wrapper scripts) |
| not supported by apps/libraries | supported by libraries like GIO (=better integrated into the app an its workflow) |
| shared xdg_{cache,config,data}_dir | private xdg_{cache,config,data}_dir |
| can sandbox privileged programs and cli programs | can not run privileged program and isn't developed for cli programs |
| same permissions for pipewire (example) | restricted permissions for pipewire (example) |
| custom seccomp filters, mdwe, big default blacklist | only a small hardcoded seccomp filter |

**Reply by CocoR55** [reply]
Flatpak uses bubblewrap under the hood for sandboxing. Bubblewrap is less complex than Firejail: https://github.com/containers/bubblewra ... n-firejail. That makes it easier to audit and maintain. On top of the things Bubblewrap sandboxes, Flatpak handles some higher level stuff such as audio while Firejail tries to handle everything in one. And flatpak is developed with Wayland in mind. I'm unclear how ready Firejail is for Wayland.

Firejail comes with default "profiles" for many popular applications so in first instance what permissions your apps have with Firejail depends on the author of the Firejail profile. Which is usually not the person that developed the app. And there may not be a profile for all your apps so you may have to write one yourself. As opposed to with a flatpak, you can't use something like Flatseal to easily limit permissions of an app—you will have to edit a configuration file by hand. And, at least this used to be the case, many Firejail profiles use blacklists—granting filesystem access to everything but what is blacklisted. Many profiles instead also use whitelisting—granting access only to that and nothing else—but the Bubblewrap developer doesn't think it's a good security practice either way.

I think in few years time my question may be moot. I think it will be become more common to use flatpaks and Flatseal makes for an easy way to tune permissions to your liking as a user. Firejail requires manually editing complex configuration files, profiles are developed out of tree from the apps they sandbox and it doesn't stack with flatpak's sandbox if that would even be useful. I can imagine for the apps you're concerned about, like web browser or things like Steam or things like PDF viewers that may view files downloaded from the internet, it will become common to use a flatpak so those more risky apps are sandboxed.

That can also be a downside. Developers are simply interested in getting stuff working and aren't really thinking about a minimalist set of restrictions that their code can work with. I say that as a developer. :lol:

I was unaware of Flatseal though, so thanks for adding that info. It's not clear from the link you providing, but is it possible to restrict access to different parts of the filesystem? What I like with firejail is being able to open up only the specific parts of my system that the program needs, e.g. the music player can see the Music directory, but not the rest of my files

**Reply by rusty-snake** [reply]
> And flatpak is developed with Wayland in mind. I'm unclear how ready Firejail is for Wayland.

I'm unaware of any special handling of Wayland by flatpak (except of `socket=wayland`). Since Wayland was already developed with isolation between clients I do not see much to do here (the only thing I miss is are clipboard permissions).

>  depends on the author of the Firejail profile

Actually I try that all the hardening commands are used if they work when a PR is opened. <sub>Not all commands are used since some require extra work/knowledge (e.g. `net enp2s0`).</sub> However profiles for uncommon programs are not updated with new hardening command for years (IMHO we need some kind of beta testers for profiles).

> many Firejail profiles use blacklists—granting filesystem access to everything but what is blacklisted. 

For some apps it is simply not possible to use whitelisting. Firejail uses a huge list of known sensitive in such cases while flatpak just `filesystem=home` it.

> but the Bubblewrap developer doesn't think it's a good security practice either way.

I never understood that because that is whitelisting (`whitelist ${HOME}/Downloads`+`whitelist ${HOME}/.ssh`):

```bash

bwrap --die-with-parent --unshare-all --new-session --cap-drop all --proc /proc --dev /dev --bind /var /var --ro-bind /etc /etc --ro-bind /usr /usr --symlink /usr/bin /bin --symlink /usr/lib /lib --symlink /usr/lib64 /lib64 --symlink /usr/sbin /sbin --tmpfs /tmp --tmpfs $HOME --bind $HOME/Downloads $HOME/Downloads --bind-try $HOME/.ssh $HOME/.ssh ls -A
Downloads  .ssh                                                                                                                                                                                                                                                        *************************************************************************************
```

> I think in few years time my question may be moot. I think it will be become more common to use flatpaks and Flatseal makes for an easy way to tune permissions to your liking as a user. Firejail requires manually editing complex configuration files, profiles are developed out of tree from the apps they sandbox and it doesn't stack with flatpak's sandbox if that would even be useful. I can imagine for the apps you're concerned about, like web browser or things like Steam or things like PDF viewers that may view files downloaded from the internet, it will become common to use a flatpak so those more risky apps are sandboxed.

Yeah I also think that firejail plays a much smaller role in a few year because daemons are sandboxed with systemd (this is already the case) and graphical apps are sandboxed with flatpak. Firejail then plays only a role for (proprietary) apps that are distributed as (tar-)archives  and cli programs.

> That can also be a downside. Developers are simply interested in getting stuff working and aren't really thinking about a minimalist set of restrictions that their code can work with.

:+1:

Maybe the community around firejail could move from writing firejail profiles to write flatpak overrides (maybe even with a lax and a strict set) and provide an updater to make it easier for users. <sub>What I miss for such a project is a proper comment support in flatpak overrides</sub>

> I was unaware of Flatseal though, so thanks for adding that info. It's not clear from the link you providing, but is it possible to restrict access to different parts of the filesystem? What I like with firejail is being able to open up only the specific parts of my system that the program needs, e.g. the music player can see the Music directory, but not the rest of my files

- `man 5 flatpak-metadata` (section `[Context]`, `[Session Bus Policy]`, `[System Bus Policy]` and `[Environment]`)
- `man 1 flatpak-override`

Simplified: A flatpak starts with no filesystem access of the host and then you add paths via `filesystem=`.

For your music player example, `flatpak info -M com.example.Music` should list only `filesystem=xdg-music;` and maybe some ignored like `!home;!host;`. If you want examples: https://github.com/rusty-snake/kyst/tree/main/flatpak.

**Reply by rusty-snake** [reply]
`flatpak update` notifies you if an updated introduces new permissions. I'm not sure if gnome-software (which I do not recommend to use for updates/installation anyway) does so. Therefore you should always use `flatpak update`.

**Reply by rusty-snake** [reply]
You maybe also want to read the firejail/flatpak related comments at https://github.com/privacyguides/privacyguides.org/discussions/169#discussioncomment-1435572.

**Reply by topimiettinen** [reply]
Somewhat related: [Flatpak Is Not the Future](https://ludocode.com/blog/flatpak-is-not-the-future), especially parts 'Security' and 'Is Flatpak Fixable?'.

**Reply by kmk3** [reply]
@topimiettinen [on Nov 25](https://github.com/netblue30/firejail/discussions/4685#discussioncomment-1699349):

> Somewhat related: [Flatpak Is Not the Future](https://ludocode.com/blog/flatpak-is-not-the-future), especially
> parts 'Security' and 'Is Flatpak Fixable?'.

Nice, I had seen this article and thought about sharing it in here as well.

It has some interesting insights about the current container tools and has
critiques that I hadn't seen before.

Though besides Flatpak it also talks about the direction of containers in
general and about native libraries.  What do you think about making a thread
specifically for the article?

**Reply by topimiettinen** [reply]
> Though besides Flatpak it also talks about the direction of containers in general and about native libraries. What do you think about making a thread specifically for the article?

Well, it's not fully related, there's no mention of Firejail at all. Many of the problems listed for flatpaks and snaps are not applicable to Firejail, maybe even the opposite (like native libraries: there's no support for replacing distro libraries with something else for an app).
